include/wil/safearrays.h

```diff
@@ -0,0 +1,889 @@
+
+        //! Retrieves the stored type (when not working with a fixed type)
+        template<typename T = element_t, typename wistd::enable_if<is_type_not_set<T>::value, int>::type = 0>
+        VARTYPE vartype() WI_NOEXCEPT
+        {
+            return details::SafeArrayGetVartype(storage_t::get());
+        }
+
+        //! Returns the current number of locks on the SAFEARRAY
+        ULONG lock_count() const WI_NOEXCEPT
+        {
+            return details::SafeArrayGetLockCount(storage_t::get());
+        }
+        //! @}
+
+        //! Returns an object that keeps a lock count on the safearray until it goes out of scope
+        //! Use to keep the SAFEARRAY from being destroyed but without accessing the contents
+        //! Not needed when using safearraydata_t because accessing the data also holds a lock
+        //! ~~~~
+        //! void Process(wil::unique_safearray& sa)
+        //! {
+        //!     auto lock = sa.scope_lock();
+        //!     LongRunningFunction(sa);
+        //! }
+        WI_NODISCARD safearray_unlock_scope_exit scope_lock() const WI_NOEXCEPT
+        {
+            return wil::SafeArrayUnlock_scope_exit(storage_t::get());
+        }
+
+        //! @name Element Access
+        //! Instead of using the safearraydata_t to access the entire collection of elements
+        //! at once, it is also possible to access them one at a time.  However, using this
+        //! functionality is less efficient because it deals with copies of the elements
+        //! that the caller is then responsible for cleaning up, so use RAII data types
+        //! for items that require clean up, like BSTR, VARIANT, or LPUNKNOWN
+        //! ~~~~
+        //! HRESULT ProcessWonderful_UsingElements()
+        //! {
+        //!     wil::unique_safearray_nothrow sa{};
+        //!     wil::unique_bstr bstr{};
+        //!     ULONG count{};
+        //!
+        //!     // Invoke the API
+        //!     RETURN_IF_FAILED(::GetWonderfulData(sa.put()));
+        //!     // Verify the output is expected
+        //!     RETURN_HR_IF(E_UNEXPECTED, sa.vartype() != VT_BSTR);
+        //!     RETURN_HR_IF(E_UNEXPECTED, sa.dims() != 1);
+        //!     RETURN_IF_FAILED(sa.count(&count));
+        //!     for(auto i = 0U; i < count; ++i)
+        //!     {
+        //!         // Copy from the safearray to the unique_bstr
+        //!         sa.get_element(i, bstr.put()); // Will release current BSTR
+        //!         DoSomethingWithBSTR(bstr);
+        //!     }
+        //!     return S_OK;
+        //! }
+        //! ~~~~
+        //! @{
+
+        //! Copies the specified element and puts it into the specified index in the SAFEARRAY (1D version)
+        //! The SAFEARRAY will own any resources and will clean up when it is destroyed
+        template<typename T = element_t, typename wistd::enable_if<is_value_type<T>::value, int>::type = 0>

```

Comment:
The default type of `T` here and below seems misleading as it'll never actually come in to play. The type of the argument will always dictate what `T` is.

Reply:
Yes, if T is provided, it will always use that T.  However, for the cases when element_t is not void, it will come into play because you don't have to provide the T at all.  I agree it is unusual to have a single template parameter change the shape of the class so dramatically.  I'm really implementing two classes:  one where the type is built-in and one where it isn't and always has to be provided when needed.

Reply:
I'm not sure I follow. `T` is the type of the second argument in the function. The caller must provide the second argument which will dictate what `T` is. The default here has no effect.

Reply:
I see what you're saying.  You're only objecting to the `= element_t` part.  It's there so that all the functions are the same and easier to visually validate.  If you think it's misleading, I'll look at removing it.

Reply:
Correct. Declaring the default suggests that you did so for a functional reason. Since the default will never have an effect, it gives the impression that you may have made a mistake. E.g. perhaps you intended for the argument to be `element_t`.